core/src/main/java/tc/oc/pgm/projectile/ProjectileMatchModule.java

```diff
@@ -88,32 +101,96 @@ && isValidProjectileAction(event.getAction(), projectileDefinition.clickAction))
 
       if (this.isCooldownActive(player, projectileDefinition)) return;
 
-      boolean realProjectile = Projectile.class.isAssignableFrom(projectileDefinition.projectile);
+      boolean realProjectile = false;
+      if (projectileDefinition.projectile instanceof ProjectileDefinition.ProjectileEntity.RealEntity) {
+        realProjectile = Projectile.class.isAssignableFrom(
+            ((ProjectileDefinition.ProjectileEntity.RealEntity) projectileDefinition.projectile).entityType
+        );
+      }
```

Comment:
given how much `projectileDefinition.projectile` is used, not just here but in the following blocks too, you may as well extract it
```suggestion
      var projType = projectileDefinition.projectile;
      boolean realProjectile = projType instanceof ProjectileDefinition.ProjectileEntity.RealEntity re 
              && Projectile.class.isAssignableFrom(re.entityType());
```
Let's make use of java 17 features too